File: backend/src/services/conciliacaoService.js
// src/services/conciliacaoService.js
// Referência bancária e conciliação automática do extrato.
//
// A PROMESSA DO PRODUTO é pagamento garantido. Até aqui a garantia era uma
// pessoa do Financeiro a abrir um PDF: funciona, e não escala. O momento em que
// deixa de escalar é o momento em que há volume — o pior momento para descobrir.
//
// A referência bancária é o caminho mais curto até uma confirmação sem toque
// humano, e não depende de contrato com gateway nenhum: cada fatura ganha uma
// referência única, o banco devolve o extrato, e as linhas casam sozinhas.
//
// A REGRA QUE GOVERNA TUDO AQUI: na dúvida, NÃO se dá por paga. Uma fatura
// marcada como paga por engano é dinheiro que o fornecedor espera e não vem, e
// a plataforma passou a mentir sobre a única coisa que promete. Uma linha que
// não casa fica a aguardar decisão humana — que é o estado em que TUDO estava
// antes, por isso não se perde nada por não adivinhar.

const prisma = require('../config/database');
const { NotFoundError, BusinessRuleError } = require('../utils/errors');
const auditService = require('./auditService');

const ESTADOS = {
  POR_CONCILIAR: 'POR_CONCILIAR',
  CONCILIADA: 'CONCILIADA',
  SEM_CORRESPONDENCIA: 'SEM_CORRESPONDENCIA',
  DIVERGENTE: 'DIVERGENTE',
};

/**
 * A referência que o pagador escreve na transferência.
 *
 * Curta e sem caracteres ambíguos de propósito: é lida em voz alta ao balcão e
 * escrita à mão num formulário do banco. O alfabeto exclui O, I, 0 e 1 — os
 * quatro que se trocam entre si e produzem uma transferência que chega ao banco
 * e não casa com fatura nenhuma. Nesse caso o dinheiro entrou, o comprador
 * jura que pagou, e ninguém consegue explicar porquê.
 */
const ALFABETO = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';

function gerarReferencia() {
  let s = '';
  const bytes = require('crypto').randomBytes(10);
  for (let i = 0; i < 10; i += 1) s += ALFABETO[bytes[i] % ALFABETO.length];
  return `KX${s.slice(0, 4)}-${s.slice(4, 8)}`;
}

/**
 * Atribui uma referência única a uma fatura.
 *
 * Repete-se em caso de colisão em vez de confiar na sorte: o espaço é grande
 * (32^8), mas "grande" não é "impossível", e uma colisão aqui seria um pagamento
 * creditado à fatura errada — o pior resultado possível deste módulo.
 */
async function atribuirReferencia(invoiceId, tx = prisma) {
  for (let tentativa = 0; tentativa < 5; tentativa += 1) {
    const referencia = gerarReferencia();
    const existe = await tx.invoice.findUnique({
      where: { referenciaPagamento: referencia }, select: { id: true },
    });
    if (existe) continue;
    await tx.invoice.update({ where: { id: invoiceId }, data: { referenciaPagamento: referencia } });
    return referencia;
  }
  throw new Error('Não foi possível gerar uma referência de pagamento única após 5 tentativas.');
}

/**
 * Extrai a referência da descrição de uma linha do extrato.
 *
 * O banco devolve a descrição como o pagador a escreveu: com espaços a mais,
 * em minúsculas, sem o hífen, com a referência no meio de outro texto. Aceitar
 * só o formato exato faria falhar a maioria das transferências reais — e cada
 * falha dessas é trabalho manual que este módulo existe para eliminar.
 */
function extrairReferencia(descricao) {
  const limpo = String(descricao || '').toUpperCase().replace(/[^A-Z0-9]/g, '');
  const m = limpo.match(/KX([ABCDEFGHJKLMNPQRSTUVWXYZ23456789]{8})/);
  return m ? `KX${m[1].slice(0, 4)}-${m[1].slice(4, 8)}` : null;
}

/**
 * Importa linhas do extrato e tenta conciliar cada uma.
 *
 * Idempotente por `idNoBanco`: reenviar o mesmo extrato é uma ocorrência banal
 * (o banco reenvia, alguém carrega duas vezes o mesmo ficheiro), e não pode
 * pagar a mesma fatura duas vezes.
 */
async function importarExtrato(linhas, actor = null) {
  const resultado = { importadas: 0, repetidas: 0, conciliadas: 0, porResolver: 0, detalhes: [] };

  for (const bruta of linhas) {
    const idNoBanco = String(bruta.idNoBanco || '').trim();
    if (!idNoBanco) throw new BusinessRuleError('Cada linha do extrato precisa de um identificador do banco.');

    const jaExiste = await prisma.linhaExtrato.findUnique({ where: { idNoBanco } });
    if (jaExiste) {
      resultado.repetidas += 1;
      continue;
    }

    const montante = Number(bruta.montante);
    const referencia = bruta.referencia || extrairReferencia(bruta.descricao);

    // Só entram entradas de dinheiro. Um débito com uma referência na descrição
    // não é um pagamento — e creditá-lo seria dar por paga uma fatura com o
    // dinheiro a SAIR.
    const credito = montante > 0;

    const criada = await prisma.linhaExtrato.create({
      data: {
        idNoBanco,
        dataValor: new Date(bruta.dataValor),
        montante,
        moeda: bruta.moeda || 'AOA',
        descricao: bruta.descricao || null,
        referencia,
        estado: ESTADOS.POR_CONCILIAR,
      },
    });
    resultado.importadas += 1;

    const desfecho = credito
      ? await tentarConciliar(criada, actor)
      : await marcar(criada.id, ESTADOS.SEM_CORRESPONDENCIA, 'Linha a débito — não é uma entrada de dinheiro.');

    if (desfecho.estado === ESTADOS.CONCILIADA) resultado.conciliadas += 1;
    else resultado.porResolver += 1;
    resultado.detalhes.push({ idNoBanco, estado: desfecho.estado, motivo: desfecho.motivo || null });
  }

  return resultado;
}

async function marcar(id, estado, motivo, invoiceId = null) {
  await prisma.linhaExtrato.update({
    where: { id },
    data: {
      estado,
      motivo,
      invoiceId,
      conciliadaEm: estado === ESTADOS.CONCILIADA ? new Date() : null,
    },
  });
  return { estado, motivo };
}

/**
 * Tenta casar UMA linha com UMA fatura.
 *
 * Três condições, e as três têm de bater. Aceitar duas em três seria transformar
 * este módulo naquilo que ele existe para evitar.
 */
async function tentarConciliar(linha, actor = null) {
  if (!linha.referencia) {
    return marcar(linha.id, ESTADOS.SEM_CORRESPONDENCIA, 'Sem referência reconhecível na descrição.');
  }

  const fatura = await prisma.invoice.findUnique({
    where: { referenciaPagamento: linha.referencia },
    include: { payment: true },
  });

  if (!fatura) {
    return marcar(linha.id, ESTADOS.SEM_CORRESPONDENCIA, `Nenhuma fatura com a referência ${linha.referencia}.`);
  }

  if (fatura.payment) {
    // Não é erro do banco nem da plataforma: é o mesmo pagamento a chegar duas
    // vezes, ou um pagamento a mais. Fica sinalizado para alguém devolver o
    // dinheiro — silenciá-lo seria ficar com ele.
    return marcar(linha.id, ESTADOS.DIVERGENTE, 'A fatura já tem pagamento registado.', fatura.id);
  }

  if (linha.moeda !== fatura.currency) {
    return marcar(linha.id, ESTADOS.DIVERGENTE,
      `Moeda diferente: extrato em ${linha.moeda}, fatura em ${fatura.currency}.`, fatura.id);
  }

  // Comparação ao cêntimo. Uma tolerância "pequena" seria uma decisão de
  // negócio disfarçada de detalhe técnico — quem aceita menos do que devia é
  // sempre o fornecedor, e não é a plataforma que pode decidir isso por ele.
  const esperado = Number(fatura.amount).toFixed(2);
  const recebido = Number(linha.montante).toFixed(2);
  if (esperado !== recebido) {
    return marcar(linha.id, ESTADOS.DIVERGENTE,
      `Valor diferente: esperado ${esperado} ${fatura.currency}, recebido ${recebido}.`, fatura.id);
  }

  // Tudo bate: paga-se, e as duas escritas vivem na mesma transação.
  await prisma.$transaction(async (tx) => {
    await tx.payment.create({
      data: {
        invoiceId: fatura.id,
        amount: fatura.amount,
        currency: fatura.currency,
        status: 'PROCESSADO',
        canal: 'REFERENCIA_BANCARIA',
        // A conciliação não tem uma pessoa por trás. Guarda-se a linha do
        // extrato como origem, em vez de atribuir a alguém que não decidiu.
        processedById: actor?.id || null,
        reference: `CONC-${linha.idNoBanco}`,
        processedAt: linha.dataValor,
      },
    });
    await tx.invoice.update({ where: { id: fatura.id }, data: { status: 'PAGA' } });
  });

  await auditService.record({
    action: 'PAGAMENTO_CONCILIADO',
    entityType: 'Invoice',
    entityId: fatura.id,
    actorId: actor?.id || null,
    actorName: actor?.name || 'Conciliação automática',
    metadata: { linhaExtrato: linha.idNoBanco, referencia: linha.referencia, montante: recebido },
  }).catch(() => {});

  return marcar(linha.id, ESTADOS.CONCILIADA, null, fatura.id);
}

/**
 * As linhas que ficaram por resolver — o trabalho que sobra para uma pessoa.
 *
 * É esta lista que diz se a conciliação está a valer a pena. Se crescer sempre,
 * o formato da descrição do banco mudou e ninguém deu por isso.
 */
async function porResolver({ page, limit } = {}) {
  const paginacao = require('../utils/paginacao');
  const p = paginacao.parametros({ page, limit });
  const where = { estado: { in: [ESTADOS.SEM_CORRESPONDENCIA, ESTADOS.DIVERGENTE] } };

  const [total, itens] = await Promise.all([
    prisma.linhaExtrato.count({ where }),
    prisma.linhaExtrato.findMany({
      where, orderBy: { dataValor: 'desc' }, skip: p.skip, take: p.take,
      include: { invoice: { select: { id: true, reference: true, amount: true, currency: true } } },
    }),
  ]);
  return paginacao.envelope(itens, total, p);
}

/** Reprocessa uma linha depois de alguém corrigir a referência à mão. */
async function reconciliarManualmente(linhaId, { referencia }, actor) {
  const linha = await prisma.linhaExtrato.findUnique({ where: { id: linhaId } });
  if (!linha) throw new NotFoundError('Linha do extrato');
  if (linha.estado === ESTADOS.CONCILIADA) {
    throw new BusinessRuleError('Esta linha já foi conciliada.');
  }
  const atualizada = await prisma.linhaExtrato.update({
    where: { id: linhaId }, data: { referencia: referencia || linha.referencia },
  });
  return tentarConciliar(atualizada, actor);
}

module.exports = {
  ESTADOS,
  gerarReferencia,
  atribuirReferencia,
  extrairReferencia,
  importarExtrato,
  tentarConciliar,
  porResolver,
  reconciliarManualmente,
};

File: backend-java/src/main/java/ao/kixima/conciliacao/ConciliacaoService.java
package ao.kixima.conciliacao;

import ao.kixima.agt.AgtSandboxSubmissionService;
import ao.kixima.audit.Actor;
import ao.kixima.audit.AuditService;
import ao.kixima.common.error.BusinessRuleException;
import ao.kixima.common.error.NotFoundException;
import ao.kixima.common.pagination.PaginaResposta;
import ao.kixima.common.pagination.Paginacao;
import ao.kixima.company.Company;
import ao.kixima.company.CompanyRepository;
import ao.kixima.conciliacao.dto.LinhaExtratoDto;
import ao.kixima.conciliacao.dto.LinhaExtratoInput;
import ao.kixima.faturacao.FaturacaoService;
import ao.kixima.invoice.Invoice;
import ao.kixima.invoice.InvoiceRepository;
import ao.kixima.invoice.InvoiceStatus;
import ao.kixima.payment.Payment;
import ao.kixima.payment.PaymentRepository;
import ao.kixima.po.PoStatus;
import ao.kixima.po.PurchaseOrder;
import ao.kixima.security.CurrentUser;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.PlatformTransactionManager;
import org.springframework.transaction.support.TransactionTemplate;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.security.SecureRandom;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Espelha backend/src/services/conciliacaoService.js — referência bancária e
 * conciliação automática do extrato. A REGRA QUE GOVERNA TUDO AQUI: na
 * dúvida, NÃO se dá por paga. Três condições (referência, moeda, valor ao
 * cêntimo) e as três têm de bater; o que não casa fica a aguardar decisão
 * humana — o estado em que tudo estava antes.
 *
 * NÃO PORTADO: o ramo {@code contract} das faturas consolidadas (call-offs).
 */
@Service
public class ConciliacaoService {

    public static final String POR_CONCILIAR = "POR_CONCILIAR";
    public static final String CONCILIADA = "CONCILIADA";
    public static final String SEM_CORRESPONDENCIA = "SEM_CORRESPONDENCIA";
    public static final String DIVERGENTE = "DIVERGENTE";

    /** Sem O, I, 0 e 1 — os quatro que se trocam entre si numa transferência escrita à mão. */
    private static final String ALFABETO = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
    private static final Pattern REFERENCIA = Pattern.compile("KX([ABCDEFGHJKLMNPQRSTUVWXYZ23456789]{8})");
    private static final SecureRandom RANDOM = new SecureRandom();

    public record Desfecho(String estado, String motivo) {
    }

    public record ResultadoImportacao(int importadas, int repetidas, int conciliadas, int porResolver, List<Map<String, Object>> detalhes) {
    }

    private final InvoiceRepository invoiceRepository;
    private final LinhaExtratoRepository linhaExtratoRepository;
    private final PaymentRepository paymentRepository;
    private final CompanyRepository companyRepository;
    private final FaturacaoService faturacaoService;
    private final AuditService auditService;
    private final AgtSandboxSubmissionService agtSandboxSubmissionService;
    private final TransactionTemplate tx;

    public ConciliacaoService(InvoiceRepository invoiceRepository, LinhaExtratoRepository linhaExtratoRepository,
                               PaymentRepository paymentRepository, CompanyRepository companyRepository,
                               FaturacaoService faturacaoService, AuditService auditService,
                               AgtSandboxSubmissionService agtSandboxSubmissionService,
                               PlatformTransactionManager transactionManager) {
        this.invoiceRepository = invoiceRepository;
        this.linhaExtratoRepository = linhaExtratoRepository;
        this.paymentRepository = paymentRepository;
        this.companyRepository = companyRepository;
        this.faturacaoService = faturacaoService;
        this.auditService = auditService;
        this.agtSandboxSubmissionService = agtSandboxSubmissionService;
        this.tx = new TransactionTemplate(transactionManager);
    }

    // --- Referência bancária ---------------------------------------------------

    static String gerarReferencia() {
        StringBuilder s = new StringBuilder();
        for (int i = 0; i < 10; i++) s.append(ALFABETO.charAt(RANDOM.nextInt(ALFABETO.length())));
        return "KX" + s.substring(0, 4) + "-" + s.substring(4, 8);
    }

    /** Atribui e persiste uma referência única na fatura já criada (gerida pela transacção do chamador); repete em caso de colisão. */
    public String atribuirReferencia(Invoice invoice) {
        for (int tentativa = 0; tentativa < 5; tentativa++) {
            String referencia = gerarReferencia();
            if (invoiceRepository.findByReferenciaPagamento(referencia).isPresent()) continue;
            invoice.setReferenciaPagamento(referencia);
            return referencia;
        }
        throw new IllegalStateException("Não foi possível gerar uma referência de pagamento única após 5 tentativas.");
    }

    /** Extrai a referência da descrição como as pessoas a escrevem — espaços, minúsculas, sem hífen, no meio de outro texto. */
    public static String extrairReferencia(String descricao) {
        String limpo = (descricao == null ? "" : descricao).toUpperCase().replaceAll("[^A-Z0-9]", "");
        Matcher m = REFERENCIA.matcher(limpo);
        return m.find() ? "KX" + m.group(1).substring(0, 4) + "-" + m.group(1).substring(4, 8) : null;
    }

    // --- Importação ------------------------------------------------------------

    private static Instant dataValor(String bruta) {
        if (bruta == null || bruta.isBlank()) throw new BusinessRuleException("Cada linha do extrato precisa da data-valor.");
        try {
            return Instant.parse(bruta);
        } catch (DateTimeParseException e) {
            try {
                return LocalDate.parse(bruta).atStartOfDay(ZoneOffset.UTC).toInstant();
            } catch (DateTimeParseException e2) {
                throw new BusinessRuleException("Data-valor inválida na linha do extrato: \"" + bruta + "\".");
            }
        }
    }

    /** Idempotente por `idNoBanco`: reenviar o mesmo extrato é banal e não pode pagar a mesma fatura duas vezes. */
    public ResultadoImportacao importarExtrato(List<LinhaExtratoInput> linhas, CurrentUser actor) {
        int importadas = 0, repetidas = 0, conciliadas = 0, porResolver = 0;
        List<Map<String, Object>> detalhes = new ArrayList<>();

        for (LinhaExtratoInput bruta : linhas) {
            String idNoBanco = bruta.idNoBanco() == null ? "" : bruta.idNoBanco().trim();
            if (idNoBanco.isEmpty()) throw new BusinessRuleException("Cada linha do extrato precisa de um identificador do banco.");

            if (linhaExtratoRepository.findByIdNoBanco(idNoBanco).isPresent()) {
                repetidas++;
                continue;
            }

            BigDecimal montante = bruta.montante() == null ? BigDecimal.ZERO : bruta.montante();
            String referencia = bruta.referencia() != null && !bruta.referencia().isBlank() ? bruta.referencia() : extrairReferencia(bruta.descricao());
            // Só entram entradas de dinheiro: um débito com referência não é um pagamento.
            boolean credito = montante.signum() > 0;

            LinhaExtrato criada = tx.execute(s -> linhaExtratoRepository.save(new LinhaExtrato(UUID.randomUUID().toString(), idNoBanco,
                    dataValor(bruta.dataValor()), montante, bruta.moeda() == null || bruta.moeda().isBlank() ? "AOA" : bruta.moeda(),
                    bruta.descricao(), referencia, Instant.now())));
            importadas++;

            Desfecho desfecho = credito
                    ? tentarConciliar(criada.getId(), actor)
                    : marcar(criada.getId(), SEM_CORRESPONDENCIA, "Linha a débito — não é uma entrada de dinheiro.", null);

            if (CONCILIADA.equals(desfecho.estado())) conciliadas++;
            else porResolver++;
            Map<String, Object> d = new LinkedHashMap<>();
            d.put("idNoBanco", idNoBanco);
            d.put("estado", desfecho.estado());
            d.put("motivo", desfecho.motivo());
            detalhes.add(d);
        }
        return new ResultadoImportacao(importadas, repetidas, conciliadas, porResolver, detalhes);
    }

    private Desfecho marcar(String linhaId, String estado, String motivo, String invoiceId) {
        tx.executeWithoutResult(s -> linhaExtratoRepository.findById(linhaId).ifPresent(l -> l.marcar(estado, motivo, invoiceId)));
        return new Desfecho(estado, motivo);
    }

    private record FaturaCandidata(String id, String reference, BigDecimal amount, String currency, String supplierCompanyId,
                                   boolean jaPaga) {
    }

    /** Tenta casar UMA linha com UMA fatura — três condições, e as três têm de bater. */
    public Desfecho tentarConciliar(String linhaId, CurrentUser actor) {
        LinhaExtrato linha = linhaExtratoRepository.findById(linhaId).orElseThrow(() -> new NotFoundException("Linha do extrato"));
        if (linha.getReferencia() == null) {
            return marcar(linhaId, SEM_CORRESPONDENCIA, "Sem referência reconhecível na descrição.", null);
        }

        FaturaCandidata fatura = tx.execute(s -> invoiceRepository.findByReferenciaPagamento(linha.getReferencia()).map(i -> {
            PurchaseOrder po = i.getPurchaseOrder();
            return new FaturaCandidata(i.getId(), i.getReference(), i.getAmount(), i.getCurrency(),
                    po == null ? null : po.getSupplierCompanyId(), paymentRepository.findByInvoiceId(i.getId()).isPresent());
        }).orElse(null));

        if (fatura == null) {
            return marcar(linhaId, SEM_CORRESPONDENCIA, "Nenhuma fatura com a referência " + linha.getReferencia() + ".", null);
        }
        if (fatura.jaPaga()) {
            // O mesmo pagamento a chegar duas vezes, ou um a mais — sinalizado para alguém devolver, nunca silenciado.
            return marcar(linhaId, DIVERGENTE, "A fatura já tem pagamento registado.", fatura.id());
        }
        if (!linha.getMoeda().equals(fatura.currency())) {
            return marcar(linhaId, DIVERGENTE, "Moeda diferente: extrato em " + linha.getMoeda() + ", fatura em " + fatura.currency() + ".", fatura.id());
        }
        // Comparação ao cêntimo — uma tolerância "pequena" seria uma decisão de negócio disfarçada de detalhe técnico.
        String esperado = fatura.amount().setScale(2, RoundingMode.HALF_UP).toPlainString();
        String recebido = linha.getMontante().setScale(2, RoundingMode.HALF_UP).toPlainString();
        if (!esperado.equals(recebido)) {
            return marcar(linhaId, DIVERGENTE, "Valor diferente: esperado " + esperado + " " + fatura.currency() + ", recebido " + recebido + ".", fatura.id());
        }

        Company supplier = fatura.supplierCompanyId() == null ? null : companyRepository.findById(fatura.supplierCompanyId()).orElse(null);

        // Tudo bate: paga-se, e as escritas vivem na mesma transação.
        Payment pagamento;
        try {
            pagamento = tx.execute(s -> {
                Invoice invoice = invoiceRepository.findById(fatura.id()).orElseThrow();
                FaturacaoService.Certificacao certificacao = faturacaoService.atribuir(linha.getDataValor(), invoice.getAmount(),
                        faturacaoService.serieReciboDoFornecedor(supplier == null ? null : supplier.getSerieFiscal()),
                        supplier == null ? null : supplier.getDataAdesaoFacturacaoElectronica());
                // A conciliação não tem uma pessoa por trás: a linha do extrato é a origem, não se atribui a quem não decidiu.
                Payment p = Payment.conciliado(UUID.randomUUID().toString(), invoice.getId(), invoice.getAmount(), invoice.getCurrency(),
                        actor == null ? null : actor.id(), "CONC-" + linha.getIdNoBanco(), linha.getDataValor(), certificacao);
                paymentRepository.saveAndFlush(p);
                invoice.setStatus(InvoiceStatus.PAGA);
                PurchaseOrder po = invoice.getPurchaseOrder();
                if (po != null) {
                    po.setStatus(PoStatus.PAGA);
                    po.setPaidAt(linha.getDataValor());
                }
                return p;
            });
        } catch (DataIntegrityViolationException duplicado) {
            // Payment.invoiceId é único — duas linhas de extrato distintas para a mesma fatura: a segunda não é falha do lote.
            return marcar(linhaId, DIVERGENTE, "A fatura já tem pagamento registado por outra linha do extrato.", fatura.id());
        }

        Map<String, Object> detail = new LinkedHashMap<>();
        detail.put("linhaExtrato", linha.getIdNoBanco());
        detail.put("referencia", linha.getReferencia());
        detail.put("montante", recebido);
        auditService.recordSafe(new AuditService.Entry(
                new Actor(actor == null ? null : actor.id(), actor == null ? "Conciliação automática" : actor.name(), null, null, null),
                "PAGAMENTO_CONCILIADO", "Invoice", fatura.id(), fatura.reference(), detail));

        if (fatura.supplierCompanyId() != null) {
            agtSandboxSubmissionService.submeter("RC", pagamento.getId(), fatura.supplierCompanyId());
        }
        return marcar(linhaId, CONCILIADA, null, fatura.id());
    }

    /** As linhas que ficaram por resolver — se esta lista só crescer, o formato do banco mudou e ninguém deu por isso. */
    public PaginaResposta<LinhaExtratoDto> porResolver(Integer page, Integer limit) {
        return tx.execute(s -> Paginacao.envelope(linhaExtratoRepository
                .findPorResolver(List.of(SEM_CORRESPONDENCIA, DIVERGENTE), Paginacao.parametros(page, limit))
                .map(LinhaExtratoDto::de)));
    }

    /** Reprocessa uma linha depois de alguém corrigir a referência à mão. */
    public Desfecho reconciliarManualmente(String linhaId, String referencia, CurrentUser actor) {
        tx.executeWithoutResult(s -> {
            LinhaExtrato linha = linhaExtratoRepository.findById(linhaId).orElseThrow(() -> new NotFoundException("Linha do extrato"));
            if (CONCILIADA.equals(linha.getEstado())) throw new BusinessRuleException("Esta linha já foi conciliada.");
            if (referencia != null && !referencia.isBlank()) linha.setReferencia(referencia);
        });
        return tentarConciliar(linhaId, actor);
    }
}
